src/hotspot/share/runtime/cpuTimeCounters.cpp

```diff
@@ -0,0 +1,130 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "runtime/cpuTimeCounters.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/os.hpp"
+
+const char* CPUTimeGroups::to_string(CPUTimeType val) {
+  switch (val) {
+    case CPUTimeType::gc_total:
+      return "total_gc_cpu_time";
+    case CPUTimeType::gc_parallel_workers:
+      return "gc_parallel_workers";
+    case CPUTimeType::gc_conc_mark:
+      return "gc_conc_mark";
+    case CPUTimeType::gc_conc_refine:
+      return "gc_conc_refine";
+    case CPUTimeType::gc_service:
+      return "gc_service";
+    case CPUTimeType::vm:
+      return "vm";
+    case CPUTimeType::conc_dedup:
+      return "conc_dedup";
+    default:
+      ShouldNotReachHere();
+  };
+}
+
+bool CPUTimeGroups::is_gc_counter(CPUTimeType val) {
+  switch (val) {
+    case CPUTimeType::gc_parallel_workers:
+      return true;
+    case CPUTimeType::gc_conc_mark:
+      return true;
+    case CPUTimeType::gc_conc_refine:
+      return true;
+    case CPUTimeType::gc_service:
+      return true;
+    default:
+      return false;
+  }
+  ShouldNotReachHere();
+}
+
+CPUTimeCounters* CPUTimeCounters::_instance = nullptr;
+
+CPUTimeCounters::CPUTimeCounters() :
+    _cpu_time_counters(),
+    _gc_total_cpu_time_diff(0) {
+  create_counter(SUN_THREADS, CPUTimeGroups::CPUTimeType::gc_total);
+}
+
+void CPUTimeCounters::inc_gc_total_cpu_time(jlong diff) {
+  Atomic::add(&_gc_total_cpu_time_diff, diff);
+}
+
+void CPUTimeCounters::publish_gc_total_cpu_time() {
+  // Ensure that we are only incrementing atomically by using Atomic::cmpxchg
+  // to set the value to zero after we obtain the new CPU time difference.
+  jlong old_value;
+  jlong fetched_value = Atomic::load(&_gc_total_cpu_time_diff);
+  jlong new_value = 0;
+  do {
+    old_value = fetched_value;
+    fetched_value = Atomic::cmpxchg(&_gc_total_cpu_time_diff, old_value, new_value);
+  } while (old_value != fetched_value);
+  get_counter(CPUTimeGroups::CPUTimeType::gc_total)->inc(fetched_value);
+}
```

Comment:
Why do we have to do this publish dance? Couldn't the closure that update the diff instead just update the counter? From what I can tell we never have multiple closures active at the same time so should be no race there?

Reply:
This two-step update does seem unnecessary, IMO.

Reply:
I agree that in the case that multiple closures are not active at the same time, we wouldn't have to implement it in this way. However, isn't it possible to have multiple closures active simultaneously, e.g.  vm thread, concurrent mark thread, concurrent refine thread? I think to account for the races there, we can only update the `_gc_total_cpu_time_diff` member variable atomically during these closure destructions, and then publish the actual updated `gc_total` counter at manually specified times via `publish_gc_total_cpu_time()`. If we were to call `publish_gc_total_cpu_time` as part of the thread closure, I believe it would be difficult to prevent races when accessing the underlying counter from the various gc-related threads.

Or maybe there is another strategy that I'm not seeing?

Reply:
Both `publish_gc_total_cpu_time` and `~ThreadTotalCPUTimeClosure` are called by the vm-thread inside a safepoint, so there shouldn't be any other threads running simultaneously, I believe.

Reply:
Me and Albert just spoke and we do see the problem that two concurrent threads could be executing the closure at the same time. So if having a total counter we need to sync the updates. But when talking we started to question how useful it is to have the gc_total counter. It is just an aggregate of the other gc-counters, but it is out of sync between safepoints. So you will always get a more accurate value by looking at the individual gc-counters.

We came to the conclusion that it would probably be easier to drop `gc_total` right now, rather than trying to keep it in sync for all updates to the individual counters. Because having them out of sync doesn't feel like a great option. 

Are we missing anything or do you agree?

Reply:
@simonis was the original suggester of this counter, so I will defer to his expertise. I do agree that dropping the counter would simplify things, but it also might not hurt to just leave it in. I'm okay with either option!

Reply:
I think the ideal approach to simplify this is to support Atomic operation on a `PerfCounter`.
We could either introduce a `PerfAtomicCounter`/`PerfAtomicLongCounter` class, or perform `Atomic::add()` on the `PerfData::_valuep` pointer. There's already `PerfData::get_address()`, so we might be able to do:

```
Atomic::add((volatile jlong *)(instance->get_counter(CPUTimeGroups::CPUTimeType::gc_total)->get_address()), net_cpu_time);
```

However, a new class `PerfAtomicCounter`  is likely cleaner. E.g., we may also want to make `PerfAtomicCounter::sample()` use a CAS. It is probably better to introduce `PerfAtomicCounter` in a separate RFE later.

Would the `Atomic::add()` with `PerfData::get_address()` approach be OK for now, or would we rather introduce a lock, or leave the `gc_total` mechanism as-is and address the out-of-sync-ness in a follow-up RFE?

IMO the out-of-sync-ness problem is minor, because users are likely to either look at a single `gc_total` counter, or look at each individual GC  CPU counter and disregard `gc_total`.

Reply:
In the interest of the RDP1 deadline, should we leave improving the sync issues with gc_total to a separate RFE? (Especially given that a "correct" design may take some time to come up with, and that gc_total being slightly out of sync is not a major issue.)

Reply:
Thanks, opened https://bugs.openjdk.org/browse/JDK-8321304.